src/hotspot/share/opto/mempointer.cpp

```diff
@@ -346,38 +481,144 @@ MemPointerAliasing MemPointerDecomposedForm::get_aliasing_with(const MemPointerD
+  return has_same_summands_as(other, 1);
 }
 
+// Examples:
+//   p1 = MemPointer[size=1, base + i + 16]
+//   p2 = MemPointer[size=1, base + i + 17]
+//   -> Always at distance 1
+//   -> p1 always adjacent and before p2 -> return true
+//
+//   p1 = MemPointer[size=4, x + y + z + 4L * i + 16]
+//   p2 = MemPointer[size=4, x + y + z + 4L * i + 20]
+//   -> Always at distance 4
+//   -> p1 always adjacent and before p2 -> return true
+//
+//   p1 = MemPointer[size=4, base1 + 4L * i1 + 16]
+//   p2 = MemPointer[size=4, base2 + 4L * i2 + 20]
+//   -> Have differing summands, distance is unknown
+//   -> Unknown if adjacent at runtime -> return false
 bool MemPointer::is_adjacent_to_and_before(const MemPointer& other) const {
-  const MemPointerDecomposedForm& s1 = decomposed_form();
-  const MemPointerDecomposedForm& s2 = other.decomposed_form();
-  const MemPointerAliasing aliasing = s1.get_aliasing_with(s2 NOT_PRODUCT( COMMA _trace ));
-  const jint size = mem()->memory_size();
-  const bool is_adjacent = aliasing.is_always_at_distance(size);
+  const MemPointerAliasing aliasing = get_aliasing_with(other NOT_PRODUCT( COMMA _trace ));
+  const bool is_adjacent = aliasing.is_always_at_distance(_size);
 
 #ifndef PRODUCT
   if (_trace.is_trace_adjacency()) {
     tty->print("Adjacent: %s, because size = %d and aliasing = ",
-               is_adjacent ? "true" : "false", size);
+               is_adjacent ? "true" : "false", _size);
     aliasing.print_on(tty);
     tty->cr();
   }
 #endif
 
   return is_adjacent;
 }
+
+// Examples:
+//   p1 = MemPointer[size=1, base + i + 16]
+//   p2 = MemPointer[size=1, base + i + 17]
+//   -> Always at distance 1
+//   -> Can never overlap -> return true
+//
+//   p1 = MemPointer[size=1, base + i + 16]
+//   p2 = MemPointer[size=1, base + i + 16]
+//   -> Always at distance 0
+//   -> Always have exact overlap -> return false
+//
+//   p1 = MemPointer[size=4, x + y + z + 4L * i + 16]
+//   p2 = MemPointer[size=4, x + y + z + 4L * i + 56]
+//   -> Always at distance 40
+//   -> Can never overlap -> return true
+//
+//   p1 = MemPointer[size=8, x + y + z + 4L * i + 16]
+//   p2 = MemPointer[size=8, x + y + z + 4L * i + 20]
+//   -> Always at distance 4
+//   -> Always have partial overlap -> return false
+//
+//   p1 = MemPointer[size=4, base1 + 4L * i1 + 16]
+//   p2 = MemPointer[size=4, base2 + 4L * i2 + 20]
+//   -> Have differing summands, distance is unknown
+//   -> Unknown if overlap at runtime -> return false
+bool MemPointer::never_overlaps_with(const MemPointer& other) const {
+  const MemPointerAliasing aliasing = get_aliasing_with(other NOT_PRODUCT( COMMA _trace ));
+
+  // The aliasing tries to compute:
+  //   distance = other - this
+  //
+  // We know that we have no overlap if we can prove:
+  //   this >= other + other.size      ||  this + this.size <= other
+  //
+  // Which we can restate as:
+  //   distance <= -other.size    ||  this.size <= distance
```

Comment:
Did you try to align it with L604?
```suggestion
  //   distance <= -other.size         ||  this.size <= distance
```